### PR TITLE
Add cohort loader

### PR DESCRIPTION
**Pull request by dankwart-de**

### CODE REVIEW

File: RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy [hunk 1]
```diff
@@ -160,15 +159,119 @@ public abstract class RuntimeService {
...
+                        if (primaryDataSet && dList)
+                            return new CohortDataSet(analysis, cohortDescription, primaryDataSet, dList)
+                        else
+                            return null;
+                }.findAll { it } as List<CohortDataSet>
+                if (cohortDatasets)
+                    return new SuperCohortDataSet(analysis, superCohortDescription, cohortDatasets)
+                else
+                    return (SuperCohortDataSet) null;
+        }.findAll { it }
```
Comment (vinjana) [inline RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy]: Dito.

File: RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy [hunk 2]
```diff
@@ -160,15 +159,119 @@ public abstract class RuntimeService {
...
+     * TODO Identify cases, where the matching mechanism should fail! Like e.g. when a dataset is missing but requested.
+     * - Using wildcards? At least one dataset should be matched for each wildcard entry.
+     * - Using no wildcards? There must be one dataset match.
+     *
+     * @param analysis
+     * @param pidFilters
+     * @param suppressInfo
+     * @return
+     */
+    List<DataSet> loadCohortDatasetsWithFilter(Analysis analysis, List<String> pidFilters, boolean suppressInfo) {
```
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy]: I'll try that

File: RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy [hunk 3]
```diff
@@ -160,15 +159,119 @@ public abstract class RuntimeService {
...
+            return [];
+        }
+
+        boolean error = false
+
+        // Checks are all done, now get the datasets..
+        datasets = pidFilters.collect {
+            String superCohortDescription ->
+                // Remove leading s[ and trailing ], split by |
+                List<CohortDataSet> cohortDatasets = superCohortDescription[2..-2].split("[|]").collect {
```
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy]: The method checks the string before. If it is invalid, this part of the code is not executed.

File: RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy [hunk 4]
```diff
@@ -160,15 +159,119 @@ public abstract class RuntimeService {
...
+                    String cohortDescription ->
+                        // Remove leading c:, split by ;
+                        String[] datasetFilters = cohortDescription[2..-1].split(StringConstants.SPLIT_SEMICOLON);
```
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy]: Don't know, why x:? c: is the prefix for a cohort.

File: RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy (hunk 2 above)
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy]: Done, hope you like it :D

File: RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy [hunk 5]
```diff
@@ -160,15 +159,119 @@ public abstract class RuntimeService {
...
+     * @param pidFilters
+     * @param suppressInfo
+     * @return
+     */
+    List<DataSet> loadCohortDatasetsWithFilter(Analysis analysis, List<String> pidFilters, boolean suppressInfo) {
+        List<DataSet> listOfDataSets = getListOfPossibleDataSets(analysis);
+        List<SuperCohortDataSet> datasets
+
+        boolean foundFaulty = false;
+        pidFilters.each {
```
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy]: I changed it a bit using the for in syntax. Seems clearer to me than .any()

File: RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy [hunk 6]
```diff
@@ -160,15 +159,119 @@ public abstract class RuntimeService {
...
+            boolean faulty = !validateDataSetLoadingString(it)
+            if (faulty)
+                logger.severe("The pid string ${it} is malformed.")
+            foundFaulty |= faulty
+        }
+        if (foundFaulty) {
+            logger.severe("The dataset list you provided contains errors, Roddy will not start jobs.")
+            return [];
+        }
```
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy]: See above

File: RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy [hunk 7]
```diff
@@ -160,15 +159,119 @@ public abstract class RuntimeService {
...
+                                if (_filter.contains("*") || _filter.contains("?")) {
+                                    if (!res) {
+                                        logger.severe("Could not find a match for cohort part: ${_filter}")
+                                        error = true;
+                                    }
+                                } else if (res.size() != 1) {
+                                    logger.severe("Only one match is allowed for cohort part: ${_filter}")
+                                    error = true;
+                                }
+                                dList += res;
```
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy]: done

File: RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy [hunk 8]
```diff
@@ -160,15 +159,119 @@ public abstract class RuntimeService {
...
+                        // Sort the list, but keep the primary set the primary set.
+                        DataSet primaryDataSet = dList[0];
+                        dList = dList.sort().unique()
+                        dList.remove(primaryDataSet)
+                        dList.add(0, primaryDataSet)
+                        if (primaryDataSet && dList)
+                            return new CohortDataSet(analysis, cohortDescription, primaryDataSet, dList)
+                        else
+                            return null;
+                }.findAll { it } as List<CohortDataSet>
```
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/core/RuntimeService.groovy]: The whole method is reduced now.

**Pull request merged**